packages/validator/src/services/attestation.ts

```diff
@@ -64,41 +65,84 @@ export class AttestationService {
     await Promise.race([sleep(this.clock.msToSlot(slot + 1 / 3), signal), this.emitter.waitForBlockSlot(slot)]);
     this.metrics?.attesterStepCallProduceAttestation.observe(this.clock.secFromSlot(slot + 1 / 3));
 
-    // Beacon node's endpoint produceAttestationData return data is not dependant on committeeIndex.
+    if (this.opts?.disableAttestationGrouping) {
+      // Attestation grouping optimization must be disabled in a distributed validator cluster as
+      // middleware clients such as charon (https://github.com/ObolNetwork/charon) expect the actual committee index
+      // when produceAttestationData endpoint is called. This is required because the middleware client itself
+      // calls produceAttestationData on the beacon node for each validator and there is a slight chance that
+      // the `beacon_block_root` (LMD GHOST vote) changes between calls which would cause a conflict between
+      // attestations submitted by Lodestar and other VCs in the cluster, resulting in aggregation failure.
+      // See https://github.com/ChainSafe/lodestar/issues/5103 for further details and references.
+      const dutiesByCommitteeIndex = groupAttDutiesByCommitteeIndex(duties);
+      await Promise.all(
+        Array.from(dutiesByCommitteeIndex.entries()).map(([index, duties]) =>
+          this.runAttestationTasksPerCommittee(duties, slot, index, signal)
+        )
+      );
+    } else {
+      // Beacon node's endpoint produceAttestationData return data is not dependant on committeeIndex.
+      // Produce a single attestation for all committees and submit unaggregated attestations in one go.
+      await this.runAttestationTasksGrouped(duties, slot, signal);
+    }
+  };
+
+  private async runAttestationTasksPerCommittee(
+    dutiesSameCommittee: AttDutyAndProof[],
+    slot: Slot,
+    index: number,
+    signal: AbortSignal
+  ): Promise<void> {
+    // Produce attestation with actual committee index
+    const attestation = await this.produceAttestation(index, slot);
+
+    // Step 1. Sign `Attestation` for each validator. Then publish all `Attestations` in one go
+    await this.signAndPublishAttestations(slot, attestation, dutiesSameCommittee);
+
+    // Step 2. after all attestations are submitted, make an aggregate.
+    // First, wait until the `aggregation_production_instant` (2/3rds of the way though the slot)
+    await sleep(this.clock.msToSlot(slot + 2 / 3), signal);
+    this.metrics?.attesterStepCallProduceAggregate.observe(this.clock.secFromSlot(slot + 2 / 3));
+
+    // Then download, sign and publish a `SignedAggregateAndProof` for each
```

Comment:
few comments are duplicated between `runAttestationTasksPerCommittee` and `runAttestationTasksGrouped` but I am not sure if further refactoring would improve clarity, and if someone looks at the code I would assume they will most of the time just look at one flow